### ISSUE TITLE
Memory leak when opening and closing a new istance of a window often

### ISSUE THREAD

**Issue by alessandro-gandolfi-ors** [AvaloniaUI/Avalonia#13152]
## Describe the bug
Opening and closing a window frequently increases usage of memory consistently, never decreasing. This is reproducible with both UsePlatformDetect() or UseHeadless(). The garbage collector doesn't free any memory even in longer periods of time. I added screenshots of the Program.cs code and the XAML of App and MainWindow. Glossario1View is a UserControl with some Grids, Textboxes, Textblocks, Borders (basically a styled table).

## To Reproduce
Put a breakpoint on the "return 0" of Main() and see the memory rising in the disgnostic tool.
The Thread.Sleep was only used to better visualize the process memory "steps".

## Expected behavior
Memory decreases after the window is closed and not used anymore.

## Screenshots
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/6d272a78-b454-4912-87ed-baa3a5e7b284)
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/b934664d-5a32-43d3-95b2-e319d4c0a229)
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/a2c38285-46a9-48be-aa2c-f33b1cda9af0)

## Environment
- OS: Windows 11 and Linux
- Avalonia-Version: .NET6 + latest nightly (11.1.999-cibuild0040497-beta)

## Additional context
Sample project
[AvaloniaMemLeak.zip](https://github.com/AvaloniaUI/Avalonia/files/12830128/AvaloniaMemLeak.zip)

**Comment by alessandro-gandolfi-ors** [comment]
Possibly related to #9662

**Comment by alessandro-gandolfi-ors** [comment]
I'm not an expert about the garbage collector, I asked to a coworker to run the project in dotMemory since I don't have a license and it seems that most of the memory is gen 0.
![gen0_memory](https://github.com/AvaloniaUI/Avalonia/assets/95684590/56fbdee9-98a5-49f7-8ca0-b6610e6d6592)

Maybe related to #10111 but in my case changing the GC mode to low latency didn't change anything...

**Comment by maxkatz6** [comment]
@alessandro-gandolfi-ors have you tried to check what holds these objects? Must be something with headless backend, as non headless platform didn’t had an obvious memory leak on plains windows.

**Comment by maxkatz6** [comment]
I don't think this pattern of headless usage is correct (nor documented). As there is no event loop in your code.

Checking in dotMemory, it confirms that window reference is kept in Dispatcher callbacks:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/3fb8678a-8f40-4d6f-9af2-88f7caf5eadd)


But since there is no event loop, there is nothing to run the dispatcher either.
Adding `Dispatcher.UIThread.RunJobs();` on each iteration removes "memory leak".
